### DISCUSSION TITLE
Dismissed modal sheet not reflected in UI image

### DISCUSSION THREAD

**Discussion by dmeehan1968** [pointfreeco/swift-snapshot-testing#535] [Q&A]
I'm testing initiating a modal sheet and dismissing it.  Other state changes in the UI are correctly captured, but for some reason the removal of the sheet doesn't happen.  I've tried using the `timeout:` parameter to add some wait time, but think there is something else causing the UI not to update.  I have checked that the state is updated correctly.

Animations are disabled in the snapshotting strategy.  This was largely copied from Episode 86 of swift-composable-architecture, but I had to make a change to how the window was obtained for iOS 15.

```swift
extension Snapshotting where Value: UIViewController, Format == UIImage {
  static var windowedImage: Snapshotting {
    return Snapshotting<UIImage, UIImage>.image.asyncPullback { vc in
      Async<UIImage> { callback in
        UIView.setAnimationsEnabled(false)

        // NB: Changed for iOS 15. UIApplication.shared.windows.first is deprecated
        guard let windowScene = UIApplication.shared.connectedScenes.first as? UIWindowScene,
              let window = windowScene.windows.first
              else { return }

        window.rootViewController = vc
        DispatchQueue.main.async {
          let image = UIGraphicsImageRenderer(bounds: window.bounds).image { ctx in
            window.drawHierarchy(in: window.bounds, afterScreenUpdates: true)
          }
          callback(image)
          UIView.setAnimationsEnabled(true)
        }
      }
    }
  }
}

class remindersTests: XCTestCase {

    func getTestData() -> [Item] {
        [
            Item(title: "First"),
            Item(title: "Second", dueDate: Calendar.current.startOfDay(for: .now)),
            Item(title: "Third", dueDate: Calendar.current.date(byAdding: .hour, value: -1, to: .now), dueDateUsesTime: true),
            Item(title: "Fourth", dueDate: Calendar.current.date(byAdding: .day, value: 2, to: .now)),
        ]
    }

    func testSnapshot() throws {

        let store = Store(
            initialState: RemindersState(
                items: getTestData()
            ),
            reducer: remindersReducer,
            environment: Void()
        )

        let viewStore = ViewStore(store)

        let vc = UIHostingController(rootView: Reminders(store: store))

        vc.view.frame = UIScreen.main.bounds

        // NB: Correctly captures initial state of screen
        assertSnapshot(matching: vc, as: .windowedImage)

        viewStore.send(.addReminder(Item(title: "")))

        // NB: Shows modal ready to edit
        assertSnapshot(matching: vc, as: .windowedImage)

        // NB: Should dismiss the modal
        viewStore.send(.addReminder(nil))

        // NB: Captured image still shows modal, even with timeout: 2
        assertSnapshot(matching: vc, as: .windowedImage)
    }

}
```

**Comment by Mika5652** [comment]
Hi @dmeehan1968,

I know your post is quiet old but did you solve your problem?

Today I tried to implement snapshot tests to prove my navigation logic and after some issues, I figured out how to do it.
I use **Xcode 15.3 (15E204a)**, **iOS 17.4** and I created a main test target to run snapshots in the host application.
I also use **TCA 1.9.2** and **SnapshotTesting 1.15.4**.

I used exact same implementation of `windowedImage` u used, except different precision on image inside:
`.image(precision: 0.95, perceptualPrecision: 0.95)` because my test sometimes failed when I tried it repeatedly for ten times in the row and I don't know why. But it's ok for me because main reason of this test is to prove that navigation works as expected.

This is how my test looks like and it works repeatedly as I expect.

```
func test_creating_measured_place_flow() {
  let store = Store(
    initialState: RootFeature.State(),
    reducer: RootFeature.init
  )

  let vc = UIHostingController(rootView: RootFeatureView(store: store))

  // 1. Show no measured place view. This will also trigger the view and the task modifier I used.
  assertSnapshot(of: vc, as: .windowedImage)

  // 2. Show new measured place view sheet
  store.send(.noMeasuredPlaces(.view(.createNewPlaceButtonTapped)))
  assertSnapshot(of: vc, as: .wait(for: 0.5, on: .windowedImage))

  // 3. Missing data alert when trying to tap add button with no changes
  store.send(.noMeasuredPlaces(.destination(.presented(.newMeasuredPlace(.view(.addButtonTapped))))))
  assertSnapshot(of: vc, as: .wait(for: 0.5, on: .windowedImage))

  // 4. Dismiss presented alert
  store.send(.noMeasuredPlaces(.destination(.presented(.newMeasuredPlace(.destination(.dismiss))))))
  assertSnapshot(of: vc, as: .wait(for: 0.5, on: .windowedImage))

  // 5. Dismiss presented new measured place view sheet
  store.send(.noMeasuredPlaces(.destination(.presented(.newMeasuredPlace(.view(.cancelButtonTapped))))))
  assertSnapshot(of: vc, as: .wait(for: 0.5, on: .windowedImage))
}
```

1. The code below number **2.** present sheet. 
2. Number **3.** and **4.** show and dismiss native alert. 
3. Number **5.** then dismiss view presented in number **2.** before.

It's worth noticing that when I try to use different `Snapshotting` strategy like `.iPhone8`, `.image` and even with `drawHierarchyInKeyWindow` enabled, it doesn't work for me. Views are not presented when I send action to Reducer, that should present sheet. Or they are presented, but not snapshoted.

I hope I helped you.

**Reply by Mika5652** [reply]
I reached the issue, where drill-down navigation using `navigationDestination` works in the snapshot test, but when I try to dismiss using `send(.destination(.dismiss))` it doesn't work.

I even see the destination is nil-ed out when I use the `printChanges` reducer but the back navigation is not performed.

When I tried to call `.destination(.dismiss)` with a little delay after the drill down, it worked as expected and the view was dismissed properly. 

Strangely, the `sheet` works well, but `navigationDestination` does not.

I will try to investigate the issue further.

**Reply by Mika5652** [reply]
After investigation, I realized I didn't wait enough time before I called dismissal action. With the little wait on the presented screen, dismiss now works.

**Comment by dmeehan1968** [comment]
Sorry it's been a while. I don't think I found a solution at the time.
